ANEXO 1. TFM GONZALO FONTÁN DE LA CRUZ

Dentro de los diferentes procesos de tokenización que existen, y que posteriormente pasaremos a desgranar, el que mejor ayuda a comprender en qué consiste este mecanismo es el que divide los fragmentos de texto en caracteres únicos.
Para poder realizarlo nosotros mismos de tal manera que un fragmento de texto inmenso podamos dividirlo en caracteres a través de Python nos vamos a servir de uno de los módulos más útiles: expresiones regulares o regular expression en inglés (re/regex). A través de expresiones regulares que son una serie de comandos que incluimos en la función podemos pedirle al ordenador que segmente las oraciones de tal manera que se divida en pequeños fragmentos. 


In [1]:
import re #Importamos la librería
txt = "Hola, esto es un texto!" #Creamos una variable con el texto que queremos trocear
resultado = re.split(r'(\s)', txt) #Esta línea la comentamos en el siguiente recuadro
print (resultado)

['Hola,', ' ', 'esto', ' ', 'es', ' ', 'un', ' ', 'texto!']


Generamos una variable en la que guardamos el resultado de dividir a través de la función split con la expresión regular <r'(\s)'> que hace al sistema buscar cualquier espacio para así dividirlo a través de split usando los espacios como criterio.

Las expresiones regulares son muy flexibles y se pueden aprender a generar sin necesidad de una máquina, pero ya que el presente proyecto se basa en explicar los largos modelos de lenguaje cabe mencionar que estos son expertos en producir regex.

El problema con este primer caso es que como se puede observar se generan elementos dentro de la variable que incluyen comas como puede ser el inicial 'Hola,', para ello tenemos que completar la expresión regular.

In [2]:
txt2 = "Hola, mundo, este es el segundo texto del trabajo"
resultado2 = re.split(r'([,.]|\s)', txt2)
print (resultado2)

['Hola', ',', '', ' ', 'mundo', ',', '', ' ', 'este', ' ', 'es', ' ', 'el', ' ', 'segundo', ' ', 'texto', ' ', 'del', ' ', 'trabajo']


Como se puede observar, en este caso la expresión regular cuenta con más elementos r'([,.]|\s)' y en este caso se incluyen también elementos como comas y puntos para que no se incluyan dentro de los elementos que hemos dividido.

Ahora nos surge un nuevo reto ya que como podemos observar los espacios se incluyen como elementos de la lista, esto se puede solucionar muy facilmente con dos opciones:

1. Iterar sobre la secuencia que hemos generado de tal manera que no incluyamos aquellos elementos que sean espacios o elementos vacíos.
2. Emplear la función .strip, un método .strip() se usa para eliminar los espacios en blanco (espacios, tabulaciones, saltos de línea) al inicio y al final de una cadena.

Ambas opciones son muy instructivas y las podemos abordar.

In [3]:
#quitar los espacios iterando
opcion1 =[]
for i in resultado2:
    if i != " " and i != "":
        opcion1.append(i)
print(opcion1)

['Hola', ',', 'mundo', ',', 'este', 'es', 'el', 'segundo', 'texto', 'del', 'trabajo']


In [4]:
#quitar los espacios a través de strip
opcion2 = [i for i in resultado2 if i.strip()]
print(opcion2)

['Hola', ',', 'mundo', ',', 'este', 'es', 'el', 'segundo', 'texto', 'del', 'trabajo']


Es interesante contemplar todas las posibilidades en forma de signos de puntuación dentro de nuestra expresión regular de tal manera que se consiga una limpieza absoluta del texto.

Para ello vamos a emplear los mismos elementos que hemos usado anteriormente:

In [5]:
txt3 = "Hola, gente--. Esto!, es un notebook explicativo."
resultado3 = re.split(r'([,.:;?_!"()\']|--|\s)', txt3) #Simplemente añadimos en la expresión regular todos esos elementos que queremos que incluya por separado
resultado3 = [i for i in resultado3 if i.strip()]
print (resultado3)

['Hola', ',', 'gente', '--', '.', 'Esto', '!', ',', 'es', 'un', 'notebook', 'explicativo', '.']


Ahora que contamos con herramientas suficientes para dividir un texto en tokens podemos probar a coger un fragmento de El Quijote para dividirlo de tal manera que consigamos tokens a raíz de él.

Para ello podemos usar varías rutas:
1. Cargar el libro en este archivo (lo haremos posteriormente)
2. Crear una variable con parte del texto como ya hemos hecho antes y trabajar sobre ella

In [6]:
quijote = """Desocupado lector, sin juramento me podrás creer que quisiera que este libro, como hijo del entendimiento, fuera el más hermoso, el más gallardo y más discreto que pudiera imaginarse. Pero no he podido yo contravenir al orden de naturaleza; que en ella cada cosa engendra su semejante. Y así, ¿qué podrá engendrar el estéril y mal cultivado ingenio mío, sino la historia de un hijo seco, avellanado, antojadizo y lleno de pensamientos varios y nunca imaginados de otro alguno, bien como quien se engendró en una cárcel, donde toda incomodidad tiene su asiento y donde todo triste ruido hace su habitación? El sosiego, el lugar apacible, la amenidad de los campos, la serenidad de los cielos, el murmurar de las fuentes, la quietud del espíritu son grande parte para que las musas más estériles se muestren fecundas y ofrezcan partos al mundo que le colmen de maravilla y de contento. Acontece tener un padre un hijo feo y sin gracia alguna, y el amor que le tiene le pone una venda en los ojos para que no vea sus faltas, antes las juzga por discreciones y lindezas y las cuenta a sus amigos por agudezas y donaires."""

In [7]:
quijotedividido = re.split(r'([,.:;?_!"()\']|--|\s)', quijote)
result = [i for i in quijotedividido if i.strip()]
print(quijotedividido)

['Desocupado', ' ', 'lector', ',', '', ' ', 'sin', ' ', 'juramento', ' ', 'me', ' ', 'podrás', ' ', 'creer', ' ', 'que', ' ', 'quisiera', ' ', 'que', ' ', 'este', ' ', 'libro', ',', '', ' ', 'como', ' ', 'hijo', ' ', 'del', ' ', 'entendimiento', ',', '', ' ', 'fuera', ' ', 'el', ' ', 'más', ' ', 'hermoso', ',', '', ' ', 'el', ' ', 'más', ' ', 'gallardo', ' ', 'y', ' ', 'más', ' ', 'discreto', ' ', 'que', ' ', 'pudiera', ' ', 'imaginarse', '.', '', ' ', 'Pero', ' ', 'no', ' ', 'he', ' ', 'podido', ' ', 'yo', ' ', 'contravenir', ' ', 'al', ' ', 'orden', ' ', 'de', ' ', 'naturaleza', ';', '', ' ', 'que', ' ', 'en', ' ', 'ella', ' ', 'cada', ' ', 'cosa', ' ', 'engendra', ' ', 'su', ' ', 'semejante', '.', '', ' ', 'Y', ' ', 'así', ',', '', ' ', '¿qué', ' ', 'podrá', ' ', 'engendrar', ' ', 'el', ' ', 'estéril', ' ', 'y', ' ', 'mal', ' ', 'cultivado', ' ', 'ingenio', ' ', 'mío', ',', '', ' ', 'sino', ' ', 'la', ' ', 'historia', ' ', 'de', ' ', 'un', ' ', 'hijo', ' ', 'seco', ',', '', ' ', 'av

Podemos observar a través de la función len() que se encarga de contabilizar cuán larga es una lista la cantidad de elementos que tenemos en nuestro fragmento dividido:

In [8]:
print(len(quijotedividido))

439


Como se puede observar estamos generando un vocabulario con un total de 439 tokens (en este caso palabras para simplificar un poco la forma en la que se comprende el formato) y esto nos puede servir para hacernos a una idea del funcionamiento del vocabulario. Ahora, una vez tenemos el vocabulario a través de crer con una lista de comprensión el vocabulario reunido en una variable que sea reconocible para nosotros (vocab) para después iterar y poder enumerarlo de modo que asociemos cada token a ese número único.

In [9]:
vocab = {token:integer for integer, token in enumerate(quijotedividido)}
for i, x in enumerate(vocab):
    if i < 50:
            print(i, x)

0 Desocupado
1  
2 lector
3 ,
4 
5 sin
6 juramento
7 me
8 podrás
9 creer
10 que
11 quisiera
12 este
13 libro
14 como
15 hijo
16 del
17 entendimiento
18 fuera
19 el
20 más
21 hermoso
22 gallardo
23 y
24 discreto
25 pudiera
26 imaginarse
27 .
28 Pero
29 no
30 he
31 podido
32 yo
33 contravenir
34 al
35 orden
36 de
37 naturaleza
38 ;
39 en
40 ella
41 cada
42 cosa
43 engendra
44 su
45 semejante
46 Y
47 así
48 ¿qué
49 podrá


Ahora vamos a instalar gensim que en resumidad cuentas es una biblioteca especializada en modelado de temas y representaciones vectoriales del lenguaje. Permite cargar modelos preentrenados como Word2Vec, FastText o GloVe, lo que facilita el análisis semántico y de similitud entre palabras, aspectos clave en lingüística computacional. A nosotros nos interesa en este caso Word2Vec para poder ver cómo se aplica la vectorización y cómo se captura la información de las palabras y sus contextos

In [15]:
!pip install gensim

   ---------------------------------------- 0.0/24.0 MB ? eta -:--:--
   -- ------------------------------------- 1.6/24.0 MB 7.6 MB/s eta 0:00:03
   ----- ---------------------------------- 3.1/24.0 MB 8.4 MB/s eta 0:00:03
   -------- ------------------------------- 5.2/24.0 MB 8.4 MB/s eta 0:00:03
   ------------ --------------------------- 7.3/24.0 MB 8.9 MB/s eta 0:00:02
   ---------------- ----------------------- 9.7/24.0 MB 9.4 MB/s eta 0:00:02
   ------------------- -------------------- 11.8/24.0 MB 9.6 MB/s eta 0:00:02
   ----------------------- ---------------- 14.2/24.0 MB 9.9 MB/s eta 0:00:01
   --------------------------- ------------ 16.5/24.0 MB 10.0 MB/s eta 0:00:01
   --------------------------- ------------ 16.8/24.0 MB 10.1 MB/s eta 0:00:01
   ----------------------------- ---------- 17.6/24.0 MB 8.5 MB/s eta 0:00:01
   -------------------------------- ------- 19.7/24.0 MB 8.6 MB/s eta 0:00:01
   ------------------------------------- -- 22.3/24.0 MB 8.9 MB/s eta 0:00:

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
blis 1.0.1 requires numpy<3.0.0,>=2.0.0, but you have numpy 1.26.4 which is incompatible.
thinc 8.3.2 requires numpy<2.1.0,>=2.0.0; python_version >= "3.9", but you have numpy 1.26.4 which is incompatible.

[notice] A new release of pip is available: 24.3.1 -> 25.1.1
[notice] To update, run: C:\Users\Gonzalo\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [16]:
import gensim.downloader as api
model = api.load("word2vec-google-news-300")

[==================================================] 100.0% 1662.8/1662.8MB downloaded


En este caso vamos a coger, una vez hemos asociado el modelo a una variable, un par de palabras de uso más común y popular (Messi en nuestro caso, ya que se puede intuir que el lector cuenta con los conocimientos como para poder comprenderlos). Los asociamos al modelo para poder ver sus vectores, pero en este caso el lector puede cambiar sus palabras a su antoja e iterar para poder ver los conceptos que desee.

In [17]:
w_vec = model
Messi = w_vec['messi']
Argentina = w_vec['argentina']

Ahora vamos a indagar el vector asociado a Messi, como se va a poder observar es un vector muy amplio (en este caso su dimensión es de 300 y se puede comprobar abriendo una celda y poniendo len(messi) dentro de ella). Esta es la longitud de dimensión de la que se habla en el documento y es la que permite que el modelo capte la semántica y las relaciones de las palabras, a mayor dimensión, mejor captación de los conceptos. Esta comprobación se puede hacer con cualquier palabra que se desee (siempre y cuando estén dentro del modelo que descarguemos, ya que están preentrenadas).

In [22]:
Messi

array([ 0.09960938, -0.04956055,  0.33007812,  0.36914062,  0.08203125,
        0.06396484,  0.0213623 , -0.22851562,  0.01000977, -0.02258301,
       -0.17773438, -0.22363281, -0.23046875, -0.03369141, -0.08984375,
        0.18847656,  0.01525879,  0.3046875 , -0.08642578, -0.1640625 ,
       -0.14941406, -0.25585938,  0.38085938,  0.09521484,  0.11962891,
       -0.09130859, -0.07421875, -0.24804688,  0.09033203,  0.13085938,
       -0.07470703,  0.13867188,  0.02868652, -0.09716797, -0.24609375,
        0.171875  , -0.1796875 ,  0.09423828,  0.17089844,  0.17089844,
       -0.01464844, -0.22753906,  0.4453125 ,  0.35742188,  0.35742188,
       -0.15722656,  0.23925781, -0.28515625, -0.3046875 ,  0.3046875 ,
       -0.10449219,  0.21679688, -0.08984375, -0.12988281, -0.02026367,
        0.17578125, -0.0201416 ,  0.08691406, -0.13867188,  0.08154297,
       -0.09375   ,  0.22558594, -0.22949219, -0.2421875 ,  0.23046875,
       -0.10742188,  0.11279297, -0.09130859,  0.24414062,  0.14

En este punto es donde vamos a poder comprender a la perfección cuál es el funcionamiento de la vectorización de forma más clara. Si se coge el vector Messi, que es el que hemos investigado más en profundidad, se le suma el vector Portugal y se le resta el vector Argentina podemos ver que la palabra con mayor similaridad que surge de dicha operación no es otra que la que podríamos esperar: Ronaldo. Este pequeño juego interactivo se puede probar con cualquier palabra, pero se recomienda tener en cuenta que este modelo previamente entrenado no cuenta con todo el vocabulario que se puede esperar, por lo que deberíamos limitar nuestras expectativas de cara a probar con ello.

In [18]:
print(w_vec.most_similar(positive=['Messi', 'Portugal'], negative=['Argentina'], topn=10))

[('Ronaldo', 0.705854594707489), ('Xavi', 0.7020642161369324), ('Iniesta', 0.6866832375526428), ('Fabregas', 0.6855268478393555), ('Drogba', 0.6765247583389282), ('Cristiano_Ronaldo', 0.6650602221488953), ("Eto'o", 0.6632802486419678), ('Ibrahimovic', 0.6594337224960327), ('Cesc_Fabregas', 0.6520246267318726), ('Puyol', 0.650626003742218)]


In [4]:
!git remote add origin https://github.com/ElDokas/Anexo-1-TFM.git


"git" no se reconoce como un comando interno o externo,
programa o archivo por lotes ejecutable.
